nbextensions/appCell2/widgets/tabs/jobStateList.js

```diff
@@ -22,52 +22,119 @@ define([
+                            initialState = null;
+                        if (i < arg.childJobs.length) {
+                            jobId = arg.childJobs[i].job_id;
+                            initialState = arg.childJobs[i].job_state;
+                        }
+                        createJobStateWidget(i, jobId, initialState, arg.clickFunction);  // can make null ones. these need to be updated.
+                    }
                 })
-                var clickFunction = arg.clickFunction;
-                return Promise.all([
-                    Object.keys(widgets).forEach((jobId) => {
-                        widgets[jobId].start({
-                            node: container.getElementsByTagName('tbody')[0],
-                            jobId: jobId,
-                            clickFunction: arg.clickFunction
-                        })
-                    })
-                ])
+                .then(() => { startParentListener() });
+            });
+        }
+
+        function startParentListener() {
+            parentListener = runtime.bus().listen({
+                channel: {
+                    jobId: parentJobId
+                },
+                key: {
+                    type: 'job-status'
+                },
+                handle: handleJobStatusUpdate
+            });
+        }
+
+        /**
+         * Pass the job state to all row widgets, if it exists.
+         * @param {Object} message
+         */
+        function handleJobStatusUpdate(message) {
+            if (message.jobState.child_jobs) {
+                message.jobState.child_jobs.forEach((state, idx) => {
+                    widgets[idx].updateState(state);
+                });
+            }
+        }
+
+        /**
+         * Creates a job state widget and puts it in the widgets object, keyed by the index.
+         * This assumes that all child job states will always be returned in the same order
+         * on each state lookup, with new ones added to the end of the list. This also adds
+         * to the bottom of the job state list table.
+         *
+         * Each job state widget knows which child job index it's in, so it can look up its
+         * state as well.
+         * @param {int} jobNumber
+         * @param {string} jobId
+         */
+        function createJobStateWidget(jobNumber, jobId, initialState, clickFunction, isParentJob) {
+            widgets[jobNumber] = JobStateListRow.make({
+                model: model
+            });
+            widgets[jobNumber].start({
+                node: createTableRow(jobNumber),
+                name: isParentJob ? 'Parent Job' : 'Child Job ' + (jobNumber+1),
+                jobId: jobId,
+                initialState: initialState,
+                isParentJob: isParentJob ? true : false,
+                clickFunction: function(jobRow, jobId, isParentJob) {
+                    Array.from(container.getElementsByTagName('tr')).forEach((elem) => {
```

Comment:
cant't you just getElementsByClassName()

Reply:
Duh. Yes. Yes I can.
:)